### ISSUE TITLE
Progress performance poor with open-loop input

### ISSUE THREAD

**Issue by frankmcsherry** [frankmcsherry/timely-dataflow#78]
When inputs are supplied with multiple distinct timestamps from workers which do not wait for the system to catch up, we can get into a spot with a large number of distinct timestamps in the system. Having a large number isn't fundamentally bad, but there are certain operations that are currently quadratic in the number of outstanding timestamps, which typically just causes overloaded workers to fall further behind.

The main culprits are 

1. `CountMap`, which just uses an unordered list for an interface that is more like a `HashMap<T, i64>`. This could be fixed, perhaps with an enum that switches the internal representation between a `Vec` and a `HashMap` based on sizes.

2. `MutableAntichain` is a bit more problematic. The "maintain the antichain lower envelope" logic is quadratic as written, and I'm not sure of a better algorithm for general partial orders. This is a problem even if the frontier is compact, as we have a large collection of non-frontier times that we have a hard time organizing (when a frontier element departs, which element next joins the frontier?).

Given these two things, it might be good to come up with a discipline for batching inputs from workers who aren't clear on how many outstanding timestamps they've introduced into the system but not yet ensured have cleared the system.

Each worker knows how many timestamps it has introduced, and by consulting its `probe`s it can determine how many of these it is sure have cleared (others may have cleared, perhaps having been dropped, but could be ahead of the probe frontier). It seems reasonable to have workers limit themselves by their own knowledge, though I am not certain if they should buffer their input indefinitely until timestamps drain, or start geometrically increasing buffer thresholds before shipping their data, or what. I'll have a ponder, unless others have thoughts too!

**Comment by frankmcsherry** [comment]
The specific example I'm looking at has each worker introduce records at a fixed rate (e.g. 1/us) and finds itself with one worker pretty free and one pretty busy. The free worker keeps jamming records in, and sending the precise progress information about timestamps, sourced in nanoseconds. 

Ideally, the worker could reflect and understand that there is a gap between the current nanosecond it is thinking about using as a timestamp, and the one most recently acknowledged as complete by its `probe`s. One simple heuristic: delay (and batch) records for as long as the current gap between input time and probe time, divided by some `k`, under the premise that adding this much additional latency is not hurting anything.

**Comment by utaal** [comment]
I think we may be encountering this, <img width="830" alt="screen shot 2017-08-14 at 14 36 20" src="https://user-images.githubusercontent.com/791335/29271982-535f277c-80fe-11e7-886d-b7f61b950914.png">

**Comment by frankmcsherry** [comment]
Sweet! There is a branch https://github.com/frankmcsherry/timely-dataflow/tree/count_map that has an initial fix checked in. It corrects the bad behavior I had produced myself, but it could be that it doesn't fix everything, or has other weird behavior.

The intent of the branch is to "batch" substantially more progress tracking work together. Right now much of the progress logic looks like:

```rust
for (time, diff) in updates.drain(..) {
    frontier.apply(time, diff);
}
```

where the `frontier.apply` method needs to maintain invariants after each call, for example correctly tracking the frontier of a bunch of accumulated counts. This means that, for example, if you have 1M messages in flight at distinct timestamps, and then you get word that all 1M were consumed, you will for each of the 1M updates tweak the frontier slightly (removing an element in the frontier) and need to rebuild, taking ~1M-ish time.

The change removes the single element update methods and exposes only batch update methods, so that you would have to write 

```rust
frontier.apply_iter(updates.drain(..));
```

which looks great, and knows that it only needs to rebuild invariants after all of the updates are accepted. It is a bit more complex than this (many uses of `update` want to see how the results changed by way of a supplied closure; it works).

This corrected the defects that I was seeing, but it will be good to get reports on whether it seems to lighten the cost of other loosely coupled systems.

**Comment by utaal** [comment]
Thanks! I'll see if I can test the branch, and will report back with a more complete perf trace.

**Comment by frankmcsherry** [comment]
This should have been fixed by #82, but I forgot there was an issue to close as well!
